heat/core/random.py

```diff
@@ -451,30 +484,48 @@ def rand(
     split = stride_tricks.sanitize_axis(shape, split)
     device = devices.sanitize_device(device)
     comm = communication.sanitize_comm(comm)
-    balanced = True
-
-    # generate the random sequence
-    if dtype == types.float32:
-        x_0, x_1, lshape, lslice = __counter_sequence(shape, torch.int32, split, device, comm)
-        x_0, x_1 = __threefry32(x_0, x_1, seed=__seed)
 
-        # combine the values into one tensor and convert them to floats
-        values = __int32_to_float32(torch.stack([x_0, x_1], dim=1).flatten()[lslice]).reshape(
-            lshape
-        )
-    elif dtype == types.float64:
-        x_0, x_1, lshape, lslice = __counter_sequence(shape, torch.int64, split, device, comm)
-        x_0, x_1 = __threefry64(x_0, x_1, seed=__seed)
+    if __rng == "Threefry":
+        # use Threefry RNG (note: batchparallel RNG comes at the very end in only line only)
+        balanced = True
+        # generate the random sequence
+        if dtype == types.float32:
+            x_0, x_1, lshape, lslice = __counter_sequence(shape, torch.int32, split, device, comm)
+            x_0, x_1 = __threefry32(x_0, x_1, seed=__seed)
+
+            # combine the values into one tensor and convert them to floats
+            values = __int32_to_float32(torch.stack([x_0, x_1], dim=1).flatten()[lslice]).reshape(
+                lshape
+            )
+        elif dtype == types.float64:
+            x_0, x_1, lshape, lslice = __counter_sequence(shape, torch.int64, split, device, comm)
+            x_0, x_1 = __threefry64(x_0, x_1, seed=__seed)
+
+            # combine the values into one tensor and convert them to floats
+            values = __int64_to_float64(torch.stack([x_0, x_1], dim=1).flatten()[lslice]).reshape(
+                lshape
+            )
+        else:
+            # Unsupported type
+            raise ValueError(f"dtype is none of ht.float32 or ht.float64 but was {dtype}")
 
-        # combine the values into one tensor and convert them to floats
-        values = __int64_to_float64(torch.stack([x_0, x_1], dim=1).flatten()[lslice]).reshape(
-            lshape
-        )
+        return DNDarray(values, shape, dtype, split, device, comm, balanced)
     else:
-        # Unsupported type
-        raise ValueError(f"dtype is none of ht.float32 or ht.float64 but was {dtype}")
-
-    return DNDarray(values, shape, dtype, split, device, comm, balanced)
+        # use batchparallel RNG
+        x = factories.__factory(
+            shape if shape != () else (1,),
+            dtype,
+            split if split is not None else 0,
+            torch.rand,
+            device,
+            comm,
+            "C",
+        )
+        if split is None:
+            x = x.resplit_(None)
+        if shape == ():
+            x = x.reshape(shape)
```

Comment:
Can the support for the special cases be integrated into __factory such that the arrays doesn't have to be changed back and forth?

Reply:
actually, I would like to avoid doing this as the reason for this workaround is not rooted in `__factories` but in the specific nature of batchparallel RNG: 

`__factories` creates a DNDarray with the local arrays being created by the input function, i.e., in our case `torch.randn`. Since torch's random seed is different (and must be different) on each MPI process for batchparallel RNG, this would result in different local arrays across processes in the case `split=None`; hence, there are three options, none of them really cool: 

*  save the states of each of the process-local torch RNGs, synchronize their states, call `__factories`, give the process-local RNGs back their unsynchronized state (I dont even know whether that works...) 

* don't use `__factories`: use torch's RNG only on one process (say process 0) and broadcast the resulting arrays to all other processes; manually generate a non-split DNDarray from these local arrays 

* tell `__factories` to create the desired array, but with artificial split=0. Then resplit. 

In my opinion, the last variant (as implemented) is the most easy one; in particular, since the second-last variant also involves communication between all processes.

Reply:
@mtar do you agree?